### PR TITLE
GROM-153: Move Session state to shared preferences

### PR DESCRIPTION
**Pull request by dan-0**
Moves session state to shared preferences and adds some optimization on code from other pending PRs.

### CODE REVIEW

File: app/src/main/java/com/rockthevote/grommet/ui/eventFlow/PartnerLoginViewModel.kt [hunk 1]
```diff
@@ -47,35 +47,31 @@ class PartnerLoginViewModel(
         } else {
             viewModelScope.launch(dispatchers.io + coroutineExceptionHandler) {
                 runCatching {
-                    rockyService.getPartnerName(partnerId.toString()).toBlocking().value()
-
-                }.onSuccess {
-                    if (it.response()!!.isSuccessful) {
```
Comment (dan-0) [inline app/src/main/java/com/rockthevote/grommet/ui/eventFlow/PartnerLoginViewModel.kt:53]: This threw an exception causing the app to hang during validation. `getPartnerName` returned a failed result, so `response` was null.

File: app/src/main/java/com/rockthevote/grommet/ui/MainActivityViewModel.kt [hunk 2]
```diff
@@ -105,14 +107,23 @@ class MainActivityViewModel(
      */
     fun loadSessionStatus() {
         viewModelScope.launch(dispatchers.io) {
-            val retainedStatus  = sessionDao.getCurrentSession()?.sessionStatus
+            val rawState = sharedPreferences.getString(KEY_SESSION_STATUS, null)
 
-            val status = retainedStatus ?: SessionStatus.PARTNER_UPDATE
+            val status = SessionStatus.fromString(rawState) ?: SessionStatus.PARTNER_UPDATE
```
Comment (dan-0) [inline app/src/main/java/com/rockthevote/grommet/ui/MainActivityViewModel.kt:112]: Should this be PARTNER_UPDATE or something else?

File: app/src/main/java/com/rockthevote/grommet/data/db/dao/SessionDao.kt [hunk 3]
```diff
@@ -15,9 +17,12 @@ interface SessionDao {
     @Query("SELECT * FROM session LIMIT 1")
     fun getCurrentSession(): Session?
 
+    @Query("SELECT * FROM session LIMIT 1")
+    fun getCurrentSessionLive(): LiveData<Session?>
+
     @Transaction
-    @Query("SELECT * FROM session")
-    fun getSessionWithRegistrations(): List<SessionWithRegistrations>
+    @Query("SELECT * FROM session LIMIT 1")
```
Comment (dan-0) [inline app/src/main/java/com/rockthevote/grommet/data/db/dao/SessionDao.kt:24]: Added Limit 1 here just to ensure results are more predictable

File: app/src/main/java/com/rockthevote/grommet/ui/MainActivity.java [hunk 4]
```diff
@@ -97,7 +97,7 @@ protected void onCreate(Bundle savedInstanceState) {
 
         viewModel = new ViewModelProvider(
                 this,
-                new MainActivityViewModelFactory(rockyService, registrationDao, sessionDao)
+                new MainActivityViewModelFactory(rockyService, registrationDao, sharedPreferences)
```
Comment (huttneab) [inline app/src/main/java/com/rockthevote/grommet/ui/MainActivity.java:100]: why not just inject the specific shared preference that we need?

File: app/src/main/java/com/rockthevote/grommet/ui/MainActivityViewModel.kt (hunk 2 above)
Comment (huttneab) [reply app/src/main/java/com/rockthevote/grommet/ui/MainActivityViewModel.kt:112]: partner update is the first step, so this is correct

File: app/src/main/java/com/rockthevote/grommet/ui/MainActivity.java (hunk 4 above)
Comment (huttneab) [reply app/src/main/java/com/rockthevote/grommet/ui/MainActivity.java:100]: actually, we only have one shared pref, this is fine.

File: app/src/main/java/com/rockthevote/grommet/ui/eventFlow/PartnerLoginViewModel.kt (hunk 1 above)
Comment (huttneab) [reply app/src/main/java/com/rockthevote/grommet/ui/eventFlow/PartnerLoginViewModel.kt:53]: 👍 good catch

File: app/src/main/java/com/rockthevote/grommet/ui/eventFlow/SessionTimeTracking.java [hunk 5]
```diff
@@ -164,17 +158,8 @@ public void onClickEdit(View v) {
...
-            new AlertDialog.Builder(getContext())
-                    .setMessage(R.string.clock_out_dialog_text)
-                    .setPositiveButton(R.string.dialog_yes, (dialogInterface, i) -> {
-                        clockOut();
-                    })
-                    .setNegativeButton(R.string.dialog_no,
-                            (dialogInterface, i) -> dialogInterface.dismiss())
-                    .create()
-                    .show();
+            viewModel.asyncCanClockOut(this::canClockOut, this::canNotClockOut);
```
Comment (huttneab) [inline app/src/main/java/com/rockthevote/grommet/ui/eventFlow/SessionTimeTracking.java:162]: this is clever

Comment (huttneab) [reply app/src/main/java/com/rockthevote/grommet/ui/eventFlow/SessionTimeTracking.java:162]: you're passing java lambda's to a kotlin higher order function, right? Didn't know you could do this.

I assume that's what the `Kotlin.Unit.Instance` lets you do, is it something similar to annotations?

**Pull request merged**